File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    void func(vector<int> nums,vector<vector<int>>& ans, vector<int>& ds, vector<int>& freq) {

        if (ds.size() == nums.size()) {
            ans.push_back(ds);
            return;
        }
        for (int i = 0; i < nums.size(); i++) {

            if (!freq[i]) {

                freq[i] = 1;
                ds.push_back(nums[i]);
                func(nums,ans,ds,freq);
                ds.pop_back();
                freq[i]=0;
            }
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> ds;
        int n = nums.size();
        vector<int> freq(n, 0);
        func( nums,  ans,ds, freq);
        return ans;
    }
};

File: 0046-permutations/0046-permutations.java
class Solution {
    public void func(int i, int[] nums,int[] vis,List<List<Integer>> ans, List<Integer> ds){

        if(ds.size()==nums.length){
            ans.add(new ArrayList<>(ds));
            return;
        }

        if (i >= nums.length) return;
        if(vis[i]==0){
            vis[i]=1;
            ds.add(nums[i]);
            func(0, nums, vis, ans, ds);
            ds.remove(ds.size()-1);
            vis[i]=0;
        }
        func(i+1, nums, vis, ans , ds);
        
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer> >  ans= new ArrayList<>();
        List<Integer> ds= new ArrayList<>();
        int[] vis= new int[nums.length];
        func(0, nums , vis , ans, ds);
        return ans;
    }
}